From 874dfc8e35cffafedb3eeb917ac8d7253323f508 Mon Sep 17 00:00:00 2001
From: Catalin Enache <catalin.enache@windriver.com>
Date: Mon, 23 Dec 2013 16:15:27 +0200
Subject: [PATCH] Don't change console attributes

Signed-off-by: Catalin Enache <catalin.enache@windriver.com>
---
 src/init.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/init.c b/src/init.c
index 27532ad..878957a 100644
--- a/src/init.c
+++ b/src/init.c
@@ -739,10 +739,10 @@ void stop_handler(int sig)
  */
 static
 void console_stty(void)
-{
+{	
 	struct termios tty;
 	int fd;
-
+#if 0
 	if ((fd = console_open(O_RDWR|O_NOCTTY)) < 0) {
 		initlog(L_VB, "can't open %s", console_dev);
 		return;
@@ -808,6 +808,7 @@ void console_stty(void)
 	(void) tcsetattr(fd, TCSANOW, &tty);
 	(void) tcflush(fd, TCIOFLUSH);
 	(void) close(fd);
+#endif	
 }
 
 /*
-- 
1.7.9.5

