From 533d2bab3dee88bf6eddec6d2d354ac2f6beca12 Mon Sep 17 00:00:00 2001
From: Serge Hallyn <serge.hallyn@ubuntu.com>
Date: Thu, 6 Aug 2015 00:34:17 -0500
Subject: [PATCH] get_map_ranges: initialize argidx to 0 at top of loop

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>
---
 libmisc/idmapping.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libmisc/idmapping.c b/libmisc/idmapping.c
index 34de3ce..0dce634 100644
--- a/libmisc/idmapping.c
+++ b/libmisc/idmapping.c
@@ -70,7 +70,7 @@ struct map_range *get_map_ranges(int ranges, int argc, char **argv)
 
 	/* Gather up the ranges from the command line */
 	mapping = mappings;
-	for (idx = 0; idx < ranges; idx++, argidx += 3, mapping++) {
+	for (idx = 0, argidx = 0; idx < ranges; idx++, argidx += 3, mapping++) {
 		if (!getulong(argv[argidx + 0], &mapping->upper)) {
 			free(mappings);
 			return NULL;
-- 
2.7.4

