--- lk_base/app/aboot/aboot.c	2014-08-07 18:47:59.363289000 +0800
+++ lk_dev/app/aboot/aboot.c	2014-12-09 14:46:03.407368000 +0800

@@ -87,6 +87,8 @@
 /* SWISTART */
 #ifdef SIERRA
 #define SQUASHFS_MAGIC_NUM 0x73717368
+/* magic number to identify UBI erase count header */
+#define UBI_MAGIC_NUM      0x23494255
 #endif /* SIERRA */
 /* SWISTOP */
 
@@ -1341,13 +1343,13 @@
 /* SWISTOP */
 		|| !strcmp(ptn->name, "recoveryfs")) {
 /* SWISTART */
-/* to add downloading of Squash FS which doesn't support MTD bad block handling */
+/* Add download support for Squash FS and UBI image types */
 #ifdef SIERRA
 		/* get magic number from image */
 		memcpy((void *)&fs_magic_num, data, sizeof(fs_magic_num));
 
-		/* squash fs has different downloading process */
-		if(fs_magic_num == SQUASHFS_MAGIC_NUM)
+		/* Squash fs and UBI image have different downloading processes */
+		if((fs_magic_num == SQUASHFS_MAGIC_NUM) || (fs_magic_num == UBI_MAGIC_NUM))
 		{
 			sz = ROUND_TO_PAGE(sz, page_mask);
 		}
