diff -uprN powerapp/start_power_config powerapp/start_power_config
--- a/start_power_config    2017-03-27 16:33:18.498330845 +0800
+++ b/start_power_config    2017-03-27 16:38:29.114481126 +0800
@@ -47,6 +47,8 @@ case "$1" in
                 echo "85 800000:90" > /sys/devices/system/cpu/cpufreq/interactive/target_loads
                 echo 40000 > /sys/devices/system/cpu/cpufreq/interactive/min_sample_time
                 echo 300000 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq
+                timeout_val=40000000000
+                echo app_init $timeout_val > /sys/power/wake_lock
                 echo mem > /sys/power/autosleep
                 ;;
         esac
