--- a/usb/compositions/sierra	2014-12-23 17:55:15.012726007 -0800
+++ b/usb/compositions/sierra	2014-12-23 17:56:08.520500903 -0800
@@ -50,7 +50,6 @@
   echo serial > /sys/class/android_usb/android$num/functions
 fi
 
-/usr/bin/ti_setswocinfo
 /usr/bin/ud_getusbinfo REM > /sys/class/android_usb/android$num/remote_wakeup
 /usr/bin/ud_getusbinfo SELF > /sys/class/android_usb/android$num/self_powered
 echo 1 > /sys/class/android_usb/android$num/enable  
