--- a/start_power_config	2017-03-28 15:03:52.100649193 +0800
+++ b/start_power_config	2017-03-28 16:54:41.336529729 +0800
@@ -46,6 +46,9 @@ case "$1" in
 		echo "1 800000:86" > /sys/devices/system/cpu/cpufreq/interactive/target_loads
 		echo 40000 > /sys/devices/system/cpu/cpufreq/interactive/min_sample_time
 		echo 400000 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq
+		timeout_val=40000000000
+		echo app_init $timeout_val > /sys/power/wake_lock
+		echo mem > /sys/power/autosleep

 		# Enable Low Power modes
 		echo 0 > /sys/module/lpm_levels/parameters/sleep_disabled
