--- a/runltp	2019-01-11 15:32:06.701021798 +0100
+++ b/runltp	2019-01-11 15:40:09.633013150 +0100
@@ -78,6 +78,7 @@
     }
     export LTPROOT=${PWD}
     export TMPBASE="/tmp"
+    export TMP_DD="${LTPROOT}/tmp"
     export PATH="${PATH}:${LTPROOT}/testcases/bin"
 
     export LTP_DEV_FS_TYPE="ext2"
@@ -536,7 +537,7 @@
 
     # Added -m 777 for tests that call tst_tmpdir() and try to
     #  write to it as user nobody
-    mkdir -m 777 -p $TMPBASE || \
+    mkdir -m 777 -p $TMPBASE $TMP_DD || \
     {
         echo "FATAL: Unable to make temporary directory $TMPBASE"
         exit 1
@@ -553,9 +554,9 @@
     # write to it as user nobody
     export TMPDIR=$TMP
 
-    chmod 777 $TMP || \
+    chmod 1777 $TMP || \
     {
-      echo "unable to chmod 777 $TMP ... aborting"
+      echo "unable to chmod 1777 $TMP ... aborting"
       exit 1
     }
 
@@ -987,7 +988,7 @@
 create_block()
 {
     #create a block device
-    dd if=/dev/zero of=${TMP}/test.img bs=1024 count=262144 >/dev/null 2>&1
+    dd if=/dev/zero of=${TMP_DD}/test.img bs=1024 count=262144 >/dev/null 2>&1
     if [ $? -ne 0 ]; then
         echo "Failed to create loopback device image, please check disk space and re-run"
         return 1
@@ -999,7 +1000,7 @@
             return 1
         else
             ##attach the created file to loop dev.
-            losetup $LOOP_DEV ${TMP}/test.img
+            losetup $LOOP_DEV ${TMP_DD}/test.img
             if [ $? -ne 0 ]; then
                 echo "losetup failed to create block device"
                 return 1
@@ -1030,7 +1031,7 @@
 cleanup()
 {
     [ "$LOOP_DEV" ] && losetup -d $LOOP_DEV
-    rm -rf ${TMP}
+    rm -rf ${TMP} ${TMP_DD}
 }
 
 
