--- a/testcases/lib/cmdlib.sh	2018-04-10 09:23:03.275991344 +0200
+++ b/testcases/lib/cmdlib.sh	2018-04-12 09:29:26.052828370 +0200
@@ -29,6 +29,24 @@
     set -x
 fi
 
+# SWISTART
+#=============================================================================
+# FUNCTION NAME:        command
+#
+# FUNCTION DESCRIPTION: Replace the built-in command from bash
+#
+# PARAMETERS:           [-v] command-to-check
+#
+# RETURNS:              0 if command exists, 1 else
+#=============================================================================
+command()
+{
+	[ $1 == "-v" ] && shift
+	which $1
+	return $?
+}
+# SWISTOP
+
 #=============================================================================
 # FUNCTION:   tst_cleanup
 # PURPOSE:    Clean up after a testcase.
--- a/runtest/ltplite	2018-04-09 16:23:46.215695760 +0200
+++ b/runtest/ltplite	2018-04-09 16:21:13.287695021 +0200
@@ -57,7 +57,8 @@
 
 
 #DESCRIPTION:Kernel system calls
-abort01 ulimit -c 1024;abort01
+#SWI change ulimit -c 1024 to 2048
+abort01 ulimit -c 2048;abort01
 
 accept01 accept01
 
@@ -66,9 +67,9 @@
 access03 access03
 access04 access04
 access05 access05
-access06 access06
+#SWI missing mkfs #access06 access06
 
-acct01 acct01
+#SWI missing mkfs #acct01 acct01
 
 adjtimex01 adjtimex01
 adjtimex02 adjtimex02
@@ -105,13 +106,13 @@
 chmod03 chmod03
 chmod04 chmod04
 chmod05 chmod05
-chmod06 chmod06
+#SWI missing mkfs #chmod06 chmod06
 chmod07 chmod07
 
 chown01 chown01
 chown02 chown02
 chown03 chown03
-chown04 chown04
+#SWI missing mkfs #chown04 chown04
 chown05 chown05
 
 chroot01 chroot01
@@ -141,7 +142,7 @@
 creat03 creat03
 creat04 creat04
 creat05 creat05
-creat06 creat06
+#SWI missing mkfs #creat06 creat06
 creat07 creat07
 creat08 creat08
 creat09 creat09
@@ -184,13 +185,13 @@
 fchmod03 fchmod03
 fchmod04 fchmod04
 fchmod05 fchmod05
-fchmod06 fchmod06
+#SWI missing mkfs #fchmod06 fchmod06
 fchmod07 fchmod07
 
 fchown01 fchown01
 fchown02 fchown02
 fchown03 fchown03
-fchown04 fchown04
+#SWI missing mkfs #fchown04 fchown04
 fchown05 fchown05
 
 fcntl01 fcntl01
@@ -269,7 +270,7 @@
 ftruncate01 ftruncate01
 ftruncate02 ftruncate02
 ftruncate03 ftruncate03
-ftruncate04 ftruncate04
+#SWI missing mkfs #ftruncate04 ftruncate04
 
 getcontext01 getcontext01
 
@@ -296,7 +297,7 @@
 getgroups01 getgroups01
 getgroups03 getgroups03
 
-gethostid01 gethostid01
+#SWI missing hostid #gethostid01 gethostid01
 
 gethostname01 gethostname01
 
@@ -370,12 +371,13 @@
 kill08 kill08
 kill09 kill09
 kill10 kill10
-kill11 ulimit -c 1024;kill11
+#SWI change ulimit -c 1024 to 2048
+kill11 ulimit -c 2048;kill11
 kill12 kill12
 
 lchown01 lchown01
 lchown02 lchown02
-lchown03 lchown03
+#SWI missing mkfs #lchown03 lchown03
 
 link01 symlink01 -T link01
 link02 link02
@@ -384,7 +386,7 @@
 link05 link05
 link06 link06
 link07 link07
-link08 link08
+#SWI missing mkfs #link08 link08
 
 listen01 listen01
 
@@ -420,7 +422,7 @@
 
 mkdir01 mkdir01
 mkdir02 mkdir02
-mkdir03 mkdir03
+#SWI missing mkfs #mkdir03 mkdir03
 mkdir04 mkdir04
 mkdir05 mkdir05
 mkdir05A symlink01 -T mkdir05
@@ -433,7 +435,7 @@
 mknod04 mknod04
 mknod05 mknod05
 mknod06 mknod06
-mknod07 mknod07
+#SWI missing mkfs #mknod07 mknod07
 mknod08 mknod08
 mknod09 mknod09
 
@@ -450,7 +452,7 @@
 mmap07 mmap07
 mmap08 mmap08
 mmap09 mmap09
-mmap16 mmap16
+#SWI missing mkfs #mmap16 mmap16
 
 modify_ldt01 modify_ldt01
 modify_ldt02 modify_ldt02
@@ -483,8 +485,8 @@
 msgctl05 msgctl05
 msgctl06 msgctl06
 msgctl07 msgctl07
-msgctl08 msgctl08
-msgctl09 msgctl09
+#SWI missing ipcs #msgctl08 msgctl08
+#SWI missing ipcs #msgctl09 msgctl09
 
 msgget01 msgget01
 msgget02 msgget02
@@ -543,10 +545,10 @@
 open05 open05
 open06 open06
 open07 open07
-open08 open08
+#SWI read-only file-system #open08 open08
 open09 open09
 open10 open10
-open11 open11
+#SWI open of dev fails. /tmp is mounted with nodev opts #open11 open11
 open12 open12
 open13 open13
 
@@ -649,13 +651,13 @@
 rename08 rename08
 rename09 rename09
 rename10 rename10
-rename11 rename11
+#SWI missing mkfs #rename11 rename11
 rename12 rename12
 rename13 rename13
 rename14 rename14
 
 rmdir01 rmdir01
-rmdir02 rmdir02
+#SWI missing mkfs #rmdir02 rmdir02
 rmdir03 rmdir03
 rmdir03A symlink01 -T rmdir03
 rmdir04 rmdir04
@@ -974,9 +976,9 @@
 #umount02 umount02 -D /dev/...
 #umount03 umount03 -D /dev/...
 
-umount2_01 umount2_01
-umount2_02 umount2_02
-umount2_03 umount2_03
+#SWI missing mkfs #umount2_01 umount2_01
+#SWI missing mkfs #umount2_02 umount2_02
+#SWI missing mkfs #umount2_03 umount2_03
 
 ustat01 ustat01
 ustat02 ustat02
@@ -987,7 +989,7 @@
 utime03 utime03
 utime04 utime04
 utime05 utime05
-utime06 utime06
+#SWI missing mkfs #utime06 utime06
 
 vfork01 vfork01
 vfork02 vfork02
@@ -1109,10 +1111,10 @@
 #gf28 growfiles -W gf28 -b -D 0 -w -g 16b -C 1 -b -i 1000 -u gfsparse-2-$$
 #gf29 growfiles -W gf29 -b -D 0 -r 1-4096 -R 0-33554432 -i 0 -L 60 -C 1 -u gfsparse-3-$$
 gf30 growfiles -W gf30 -D 0 -b -i 0 -L 60 -u -B 1000b -e 1 -o O_RDWR,O_CREAT,O_SYNC -g 20480 -T 10 -t 20480 gf-sync-$$
-rwtest01 export LTPROOT; rwtest -N rwtest01 -c -q -i 60s  -f sync 10%25000:rw-sync-$$
-rwtest02 export LTPROOT; rwtest -N rwtest02 -c -q -i 60s  -f buffered 10%25000:rw-buffered-$$
-rwtest03 export LTPROOT; rwtest -N rwtest03 -c -q -i 60s -n 2  -f buffered -s mmread,mmwrite -m random -Dv 10%25000:mm-buff-$$
-rwtest04 export LTPROOT; rwtest -N rwtest04 -c -q -i 60s -n 2  -f sync -s mmread,mmwrite -m random -Dv 10%25000:mm-sync-$$
-rwtest05 export LTPROOT; rwtest -N rwtest05 -c -q -i 50 -T 64b 500b:/tmp/rwtest01%f
+#SWI missing bash #rwtest01 export LTPROOT; rwtest -N rwtest01 -c -q -i 60s  -f sync 10%25000:rw-sync-$$
+#SWI missing bash #rwtest02 export LTPROOT; rwtest -N rwtest02 -c -q -i 60s  -f buffered 10%25000:rw-buffered-$$
+#SWI missing bash #rwtest03 export LTPROOT; rwtest -N rwtest03 -c -q -i 60s -n 2  -f buffered -s mmread,mmwrite -m random -Dv 10%25000:mm-buff-$$
+#SWI missing bash #rwtest04 export LTPROOT; rwtest -N rwtest04 -c -q -i 60s -n 2  -f sync -s mmread,mmwrite -m random -Dv 10%25000:mm-sync-$$
+#SWI missing bash #rwtest05 export LTPROOT; rwtest -N rwtest05 -c -q -i 50 -T 64b 500b:/tmp/rwtest01%f
 
 
