--- a/start_power_config	2016-06-20 15:14:37.965304500 +0800
+++ b/start_power_config	2016-10-10 11:22:16.524252952 +0800
@@ -47,6 +47,10 @@ case "$1" in
 		echo 40000 > /sys/devices/system/cpu/cpufreq/interactive/min_sample_time
 		echo 400000 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq
 
+		echo mem > /sys/power/autosleep
+		timeout_val=40000000000
+		echo app_init $timeout_val > /sys/power/wake_lock
+
 		# Enable Low Power modes
 		echo 0 > /sys/module/lpm_levels/parameters/sleep_disabled
 		;;
