diff -uNr a/libalsa-intf/aplay.c b/libalsa-intf/aplay.c
--- a/libalsa-intf/aplay.c	2016-06-22 09:29:19.360724398 +0800
+++ b/libalsa-intf/aplay.c	2016-11-04 12:30:40.833505764 +0800
@@ -316,7 +316,14 @@
           pcm_close(pcm);
           return -errno;
        }
-        while(1);
+/* SWISTART */
+#ifndef SIERRA
+              while(1);
+#else
+              while(1)
+                sleep(1);
+#endif
+/* SWISTOP */
     }
 
     remainingData = data_sz;
diff -uNr a/libalsa-intf/arec.c b/libalsa-intf/arec.c
--- a/libalsa-intf/arec.c	2016-06-22 09:29:19.360724398 +0800
+++ b/libalsa-intf/arec.c	2016-11-04 12:30:31.196849764 +0800
@@ -268,7 +268,14 @@
             pcm_close(pcm);
             return -errno;
 	}
-        while(1);
+/* SWISTART */
+#ifndef SIERRA
+      while(1);
+#else
+      while(1)
+        sleep(1);
+#endif
+/* SWISTOP */
    }
 
     if (flags & PCM_MMAP) {
