From bbe51c099c8823d45662e212b9b62ecbf585799c Mon Sep 17 00:00:00 2001
From: Bertrand Roussel <broussel@sierrawireless.com>
Date: Tue, 23 May 2017 16:58:11 -0700
Subject: [PATCH] Use yocto LDFLAGS / CFLAGS / gcc

---
 yaffs2/utils/Makefile | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/utils/Makefile b/utils/Makefile
index cdd41a7..a1b21de 100644
--- a/utils/Makefile
+++ b/utils/Makefile
@@ -16,15 +16,13 @@
 
 #KERNELDIR = /usr/src/kernel-headers-2.4.18
 
-CFLAGS =   -I/usr/include -I.. -O2 -Wall -DCONFIG_YAFFS_UTIL
+CFLAGS+=   -I.. -O2 -Wall -DCONFIG_YAFFS_UTIL
 CFLAGS+=   -Wshadow -Wpointer-arith -Wwrite-strings -Wstrict-prototypes -Wmissing-declarations
 CFLAGS+=   -Wmissing-prototypes -Wredundant-decls -Wnested-externs -Winline
 
 ## Change if you are using a cross-compiler
 MAKETOOLS = 
 
-CC=$(MAKETOOLS)gcc
-
 COMMONLINKS = yaffs_ecc.c
 COMMONOBJS = $(COMMONLINKS:.c=.o)
 
@@ -44,10 +42,10 @@ $(COMMONOBJS) $(MKYAFFSIMAGEOBJS) $(MKYAFFS2IMAGEOBJS) : %.o: %.c
 	$(CC) -c $(CFLAGS) $< -o $@
 
 mkyaffsimage: $(COMMONOBJS) $(MKYAFFSIMAGEOBJS)
-	$(CC) -o $@ $(COMMONOBJS) $(MKYAFFSIMAGEOBJS)
+	$(CC) $(LDFLAGS) -o $@ $(COMMONOBJS) $(MKYAFFSIMAGEOBJS)
 
 mkyaffs2image: $(COMMONOBJS) $(MKYAFFS2IMAGEOBJS)
-	$(CC) -o $@ $(COMMONOBJS) $(MKYAFFS2IMAGEOBJS)
+	$(CC) $(LDFLAGS) -o $@ $(COMMONOBJS) $(MKYAFFS2IMAGEOBJS)
 
 
 clean:
-- 
2.1.4

