diff --git a/rmnetctl/inc/librmnetctl.h b/rmnetctl/inc/librmnetctl.h
index ff78011..9f6437c 100644
--- a/rmnetctl/inc/librmnetctl.h
+++ b/rmnetctl/inc/librmnetctl.h
@@ -317,10 +317,10 @@ int rmnet_get_link_ingress_data_format_tailspace(rmnetctl_hndl_t *hndl,
 						uint8_t  *tail_spacing,
 						uint16_t *error_code);
 
-inline int rmnet_set_link_ingress_data_format(rmnetctl_hndl_t *hndl,
-					      uint32_t ingress_flags,
-					      const char *dev_name,
-					      uint16_t *error_code)
+static inline int rmnet_set_link_ingress_data_format(rmnetctl_hndl_t *hndl,
+						     uint32_t ingress_flags,
+						     const char *dev_name,
+						     uint16_t *error_code)
 {
 	return rmnet_set_link_ingress_data_format_tailspace(hndl,
 							    ingress_flags,
@@ -329,10 +329,10 @@ inline int rmnet_set_link_ingress_data_format(rmnetctl_hndl_t *hndl,
 							    error_code);
 }
 
-inline int rmnet_get_link_ingress_data_format(rmnetctl_hndl_t *hndl,
-					      const char *dev_name,
-					      uint32_t *ingress_flags,
-					      uint16_t *error_code)
+static inline int rmnet_get_link_ingress_data_format(rmnetctl_hndl_t *hndl,
+						     const char *dev_name,
+						     uint32_t *ingress_flags,
+						     uint16_t *error_code)
 {
 	return rmnet_get_link_ingress_data_format_tailspace(hndl,
 							    dev_name,
