--- a/Makefile	2016-08-25 09:37:50.000000000 -0700
+++ b/Makefile	2016-08-25 18:20:47.587013370 -0700
@@ -6,7 +6,7 @@
 includedir = $(prefix)/include
 pkgconfigdir = $(libdir)/pkgconfig
 
-CFLAGS = -Wall -Wextra
+CFLAGS ?= -Wall -Wextra -fPIC
 LDFLAGS_GCSECTIONS = -Wl,--gc-sections
-LDFLAGS = $(if $(gc_sections-pass),$(LDFLAGS_GCSECTIONS))
+LDFLAGS += $(if $(gc_sections-pass),$(LDFLAGS_GCSECTIONS))
 
@@ -35,6 +35,9 @@
 
 INSTALL_TARGETS += $(bindir)/cbordump
 INSTALL_TARGETS += $(libdir)/libtinycbor.a
+INSTALL_TARGETS += $(libdir)/libtinycbor.so
+INSTALL_TARGETS += $(libdir)/libtinycbor.so.0
+INSTALL_TARGETS += $(libdir)/libtinycbor.so.0.2
 INSTALL_TARGETS += $(pkgconfigdir)/tinycbor.pc
 INSTALL_TARGETS += $(TINYCBOR_HEADERS:src/%=$(includedir)/tinycbor/%)
 
@@ -85,7 +86,7 @@
 endif
 
 # Rules
-all: .config lib/libtinycbor.a bin/cbordump tinycbor.pc
+all: .config lib/libtinycbor.a lib/libtinycbor.so bin/cbordump tinycbor.pc
 all: $(if $(JSON2CBOR_SOURCES),bin/json2cbor)
 check: tests/Makefile | lib/libtinycbor.a
 	$(MAKE) -C tests check
@@ -101,6 +102,10 @@
 lib/libtinycbor.a: $(TINYCBOR_SOURCES:.c=.o) | lib
 	$(AR) cqs $@ $^
 
+lib/libtinycbor.so: $(TINYCBOR_SOURCES:.c=.o)
+	$(CC) -shared -Wl,-soname,libtinycbor.so.0 -o lib/libtinycbor.so.0.2 $(LDFLAGS) $^
+	cd lib ; ln -s libtinycbor.so.0.2 libtinycbor.so ; ln -s libtinycbor.so.0.2 libtinycbor.so.0
+
 bin/cbordump: $(CBORDUMP_SOURCES:.c=.o) lib/libtinycbor.a | bin
 	$(CC) -o $@ $(LDFLAGS) $^ $(LDLIBS) -lm
 
@@ -152,6 +157,7 @@
 	$(RM) bin/json2cbor
 	$(RM) lib/libtinycbor.a
 	$(RM) tinycbor.pc
+	$(RM) lib/libtinycbor.so*
 	test -e tests/Makefile && $(MAKE) -C tests clean || :
 
 distclean: clean
