diff --git a/asoc/codecs/wcd9xxx-mbhc.c b/asoc/codecs/wcd9xxx-mbhc.c
index 8ab4e707..1de7fb02 100644
--- a/asoc/codecs/wcd9xxx-mbhc.c
+++ b/asoc/codecs/wcd9xxx-mbhc.c
@@ -399,8 +399,9 @@ static bool __wcd9xxx_switch_micbias(struct wcd9xxx_mbhc *mbhc,
 		pr_debug("%s: VDDIO switch enabled\n", __func__);
 	} else if (!vddio_switch && mbhc->mbhc_micbias_switched) {
 		if ((!checkpolling || mbhc->polling_active) &&
-		    restartpolling)
+		    restartpolling) {
 			wcd9xxx_pause_hs_polling(mbhc);
+		}
 
 			snd_soc_update_bits(codec, WCD9XXX_A_MAD_ANA_CTRL,
 					    0x10, 0x10);
