From 77e95a5696618f17fca9c9f4765c5eaa863743fa Mon Sep 17 00:00:00 2001
From: poonam infosys <pinfosys@sierrawireless.com>
Date: Thu, 8 Jun 2017 13:47:03 -0700
Subject: [PATCH] 0002-QTI9X07-125-Do-not-start-adb-if-not-enabled

---
 adb/start_adbd | 40 ++++++++++++++++++++++------------------
 1 file changed, 22 insertions(+), 18 deletions(-)

diff --git a/adb/start_adbd b/adb/start_adbd
index 39ad7b9..25f31b9 100644
--- a/adb/start_adbd
+++ b/adb/start_adbd
@@ -37,31 +37,35 @@ fi

 case "$1" in
   start)
-        echo -n "Starting adbd: "
-        start-stop-daemon -S -b -a /sbin/adbd
-        # Check first if the /etc/adb_devid already has an id saved from earlier boot up.
-        # If yes use it to retrieve adb devices else use the msm ID and save it to /etc/adb_devid for subsequent boot ups.
+        compstr=`grep -w ffs /sys/class/android_usb/android0/functions || true`
+        if [ ! -z "$compstr" ] ; then
+                echo -n "Starting adbd: "
+                start-stop-daemon -S -b -a /sbin/adbd
+                # Check first if the /etc/adb_devid already has an id saved from earlier boot up.
+                # If yes use it to retrieve adb devices else use the msm ID and save it to /etc/adb_devid for subsequent boot ups.

-        if [ -e "/etc/adb_devid" ]
-        then
-                cat /etc/adb_devid > /sys/class/android_usb/android0/iSerial
-        else
-                ret_val=`cat /proc/cmdline | grep "androidboot.serialno" |wc -l`
-                if  [ $ret_val -eq 1 ]
+                if [ -e "/etc/adb_devid" ]
                 then
-                    serial_no=`cat /proc/cmdline | awk '{ for ( n=1; n<=NF; n++ ) if($n ~ "androidboot.serial") print $n }' | awk '{split($0,a, "=");print a[2]}'`
-                    echo $serial_no > /sys/class/android_usb/android0/iSerial
+                        cat /etc/adb_devid > /sys/class/android_usb/android0/iSerial
                 else
-                    cat /proc/sys/kernel/random/boot_id | cut -d- -f5 > /sys/class/android_usb/android0/iSerial
+                        serial_no=`egrep -o "androidboot\.serialno=[0-9a-fA-F]+" /proc/cmdline |cut -f2 -d=`
+                        if [ ! -z ${serial_no} ] ; then
+                            echo $serial_no > /sys/class/android_usb/android0/iSerial
+                        else
+                            cat /proc/sys/kernel/random/boot_id | cut -d- -f5 > /sys/class/android_usb/android0/iSerial
+                        fi
+                            cat /sys/class/android_usb/android0/iSerial > /etc/adb_devid
                 fi
-                    cat /sys/class/android_usb/android0/iSerial > /etc/adb_devid
+                echo "done"
         fi
-        echo "done"
         ;;
   stop)
-        echo -n "Stopping adbd: "
-        start-stop-daemon -K -n adbd
-        echo "done"
+        compstr=`grep -w ffs /sys/class/android_usb/android0/functions || true`
+        if [ ! -z "$compstr" ] ; then
+                echo -n "Stopping adbd: "
+                start-stop-daemon -K -n adbd
+                echo "done"
+        fi
         ;;
   restart)
         $0 stop
--
1.9.1

