From c01d063b91d4fae3501bd76e3a13346e01f54539 Mon Sep 17 00:00:00 2001
From: Lucy He <LHe@sierrawireless.com>
Date: Mon, 27 Nov 2017 22:35:22 +0800
Subject: [PATCH] start-power-config

Signed-off-by: Lucy He <LHe@sierrawireless.com>
---
 powerapp/start_power_config | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/start_power_config b/start_power_config
index 43b10a2..c7a45fb 100755
--- a/start_power_config
+++ b/start_power_config
@@ -45,6 +45,8 @@ case "$1" in
 		echo 998400 > /sys/devices/system/cpu/cpufreq/interactive/hispeed_freq
 		echo "1 806400:80" > /sys/devices/system/cpu/cpufreq/interactive/target_loads
 		echo 40000 > /sys/devices/system/cpu/cpufreq/interactive/min_sample_time
+		timeout_val=40000000000
+		echo app_init $timeout_val > /sys/power/wake_lock
 		echo 806400 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq
 		echo mem > /sys/power/autosleep
 
-- 
1.9.1

