Index: popt-1.16/poptconfig.c
===================================================================
--- popt-1.16.orig/poptconfig.c
+++ popt-1.16/poptconfig.c
@@ -24,6 +24,8 @@ extern int fnmatch (const char *__patter
 #if defined(HAVE_GLOB_H)
 #include <glob.h>
 
+__asm__(".symver glob,glob@GLIBC_2.2.5");
+
 #if defined(__LCLINT__)
 /*@-declundef -exportheader -incondefs -protoparammatch -redecl -type @*/
 extern int glob (const char *__pattern, int __flags,
