Index: git/src/wrap-file.c
===================================================================
--- git.orig/src/wrap-file.c
+++ git/src/wrap-file.c
@@ -41,6 +41,8 @@
 #define MAXSYMLINKS 20
 #endif
 
+__asm__(".symver glob,glob@GLIBC_2.2.5");
+
 extern char *canon_filename (const char *name, int nested, struct stat64 *stp,
 			     const char *chroot, int allow_last_link,
 			     int allow_missing);
